Updates Fungible Token tutorial to use a different structure and explainations

joshuahannan opened the pull request:
Closes https://github.com/onflow/cadence/issues/1192

## Description

* Updates the fungible token tutorial to explain the basic version of fungible tokens first before adding interfaces and capabilities
* Adds more clarification about certain topics and warnings about anti-patters
* Fixes a few sentences in the hello world and nft tutorials

______

<!-- Complete: -->

- [x] Targeted PR against `master` branch
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

MaxStalker reviewed (approved):
Looking good! :)

10thfloor commented on docs/tutorial/02-hello-world.mdx:
```diff
@@ -84,11 +84,12 @@ of them conflicting with other projects' identifiers.
+- The `private` domain is like a private API. You can optionally create [links](/cadence/language/capability-based-access-control/)
+  to any of your stored assets here.
+  Only the owner and anyone they give access to can use these interfaces to call functions that are defined in their stored assets.
+- The `public` domain is kind of like your account's public API. It is exactly like the `private` domain in that
```
```suggestion
- The `public` domain is your account's public API. It is exactly like the `private` domain in that
```

10thfloor commented on docs/tutorial/02-hello-world.mdx:
```diff
@@ -84,11 +84,12 @@ of them conflicting with other projects' identifiers.
+  Only the owner and anyone they give access to can use these interfaces to call functions that are defined in their stored assets.
+- The `public` domain is kind of like your account's public API. It is exactly like the `private` domain in that
+  the owner can link capabilities here that point to stored assets. 
+  The difference is that anyone in the network can access the `public` domain's functionality.
```
```suggestion
  Anyone in the network can access the `public` domain's functionality.
```

10thfloor commented on docs/tutorial/02-hello-world.mdx:
```diff
@@ -242,22 +243,9 @@ are available within your transaction by using the `ContractName.member` syntax.
-You can have multiple signers of a transaction by clicking multiple account avatars in the playground,
-but the number of parameters of the prepare block of the transaction
-NEEDS to be the same as the number of signers. If not, this will cause an error.
+1. The `prepare` phase is required, but is not used here, so it will be covered later.
```
```suggestion
1. The `prepare` phase is required, but is not used here. We'll cover this phase later on.
```

10thfloor commented on docs/tutorial/02-hello-world.mdx:
```diff
@@ -338,7 +326,8 @@ interact with other objects of those types.
 This contract that we just deployed declares a definition for the `HelloAsset` resource.
 
-Let's walk through this contract:
+Let's walk through this contract, starting with the resource, one of the most important things
```
```suggestion
Let's walk through this contract, starting with the resource. Resources are one of the most important things
```

10thfloor commented on docs/tutorial/02-hello-world.mdx:
```diff
@@ -349,16 +338,19 @@ pub resource HelloAsset {
-Structs from other conventional programming languages are not an ideal way to represent this ownership because they can be copied.
+Resources are useful when you want to model direct ownership.
+Traditional structs or classes from other conventional programming languages
+are not an ideal way to represent this ownership because they can be copied.
```
```suggestion
are not an ideal way to represent ownership because they can be *copied*.
```

10thfloor commented on docs/tutorial/02-hello-world.mdx:
```diff
@@ -514,6 +506,25 @@ transaction {
+This transaction also is our first to use the `prepare` phase! 
+The `prepare` phase is the only place that has access to the signing accounts' [private `AuthAccount` object](/cadence/language/accounts/#authaccount).
+`AuthAccount` has special methods that allow saving to and loading from `/storage/`,
+and creating `/private/` and `/public/` links to the objects in `/storage/`.
+(called [capabilities](/cadence/language/capability-based-access-control), will be covered later)
```
```suggestion
and creating `/private/` and `/public/` links to the objects in `/storage/`, called [capabilities](/cadence/language/capability-based-access-control) (more on these later).
```

10thfloor commented on docs/tutorial/02-hello-world.mdx:
```diff
@@ -514,6 +506,25 @@ transaction {
+By not allowing the execute phase to access account storage,
+we can statically verify which assets and areas of the signers storage a given transaction can modify.
+Browser wallets and applications that submit transactions for users can use this to show what a transaction could alter,
+and users can have more confidence that they aren't getting fed
```
```suggestion
giving users information about transactions that wallets will be executing for them, and confidence that they aren't getting fed
```

10thfloor replied:
Meh, something didn't feel right here, but take-or-leave this one imo.

turbolent reviewed (approved):
Nice!

rheaplex commented on docs/tutorial/02-hello-world.mdx:
```diff
@@ -84,11 +84,12 @@ of them conflicting with other projects' identifiers.
+- The `private` domain is like a private API. You can optionally create [links](/cadence/language/capability-based-access-control/)
+  to any of your stored assets here.
+  Only the owner and anyone they give access to can use these interfaces to call functions that are defined in their stored assets.
+- The `public` domain is your account's public API. It is exactly like the `private` domain in that
+  the owner can link capabilities here that point to stored assets.
+  Anyone in the network can access the `public` domain's functionality.
```
```suggestion
- The `public` domain is your account's public API. It is like the `private` domain in that
  only the owner can link capabilities here that point to stored assets.
  The difference is that anyone in the network can access the `public` domain's functionality.
```

rheaplex commented on docs/tutorial/03-fungible-tokens.mdx:
````diff
@@ -28,13 +28,39 @@ In this tutorial, we're going to deploy, store, and transfer fungible tokens.
+```cadence
+contract LedgerToken {
+    // Tracks every user's balance
+    pub let balances: {Address: UFix64}
````
I know this is "don't do this code" but this dictionary is publicly mutatable. Maybe  make it`access(contract) let` ?

joshuahannan replied:
Good point. I'll change that

rheaplex commented on docs/tutorial/03-fungible-tokens.mdx:
```diff
@@ -73,11 +97,13 @@ in your account and transfer to and accept tokens from other users.
 </Callout>
 
-The concepts involved in implementing a fungible token in Cadence can be hard to grasp at first. For an in-depth explanation of this functionality and code, continue reading the next section.
+The concepts involved in implementing a fungible token in Cadence can be hard to grasp at first.
```
```suggestion
The concepts involved in implementing a fungible token in Cadence can be unfamiliar at first.
```

rheaplex commented on docs/tutorial/03-fungible-tokens.mdx:
```diff
@@ -152,8 +184,8 @@ The `ExampleToken` contract only needs to be stored in the initial account that
-    // we use unsigned integers for balances because they do not require the
-    // concept of a negative number
+    // we use unsigned fixed point numbers for balances 
+    // because they can represent decimals and do not require the concept of a negative number
```
```suggestion
    // because they can represent decimals and do not allow negative values
```

rheaplex commented on docs/tutorial/03-fungible-tokens.mdx:
```diff
@@ -197,20 +237,23 @@ pub fun deposit(from: @Vault) {
 A resource can never be left hanging in a piece of code. It either needs to be explicitly destroyed or stored in an account's storage.
 
-When interacting with resources, you use the `@` symbol and a special “move operator” `<-`.
+When interacting with resources, you use the `@` symbol to specify the type and a special “move operator” `<-` in assignments.
```
👍

rheaplex commented on docs/tutorial/03-fungible-tokens.mdx:
```diff
@@ -228,90 +271,34 @@ self.balance = self.balance - amount
 
 In Solidity, this could be a risk for integer overflow or underflow, but Cadence has built-in overflow and underflow protection, so it is not a risk.
-We are also using unsigned integers in this example, so the vault`s balance cannot go below 0.
+We are also using unsigned numbers in this example, so the vault`s balance cannot go below 0.
```
We mention this advantage above, so if we wish to re-mention it we could say "As we mentioned above...".

rheaplex commented on docs/tutorial/03-fungible-tokens.mdx:
```diff
@@ -512,22 +430,200 @@ Account storage is indexed with paths, which consist of a domain and identifier.
+which is a much more costly interaction.
 
-You are now ready to run transactions that use the fungible tokens!
+In the real world, you'll likely be transferring tokens to other accounts.
```
```suggestion
In production code, you'll likely be transferring tokens to other accounts.
```

rheaplex commented on docs/tutorial/03-fungible-tokens.mdx:
```diff
@@ -512,22 +430,200 @@ Account storage is indexed with paths, which consist of a domain and identifier.
+---
+
+Another important feature in Cadence is its utilization of **Capability Security.**
+This feature ensures that, while the withdraw function is public,
```
```suggestion
This feature ensures that, while the withdraw function is declared public on the resource,
```

rheaplex commented on docs/tutorial/03-fungible-tokens.mdx:
```diff
@@ -512,22 +430,200 @@ Account storage is indexed with paths, which consist of a domain and identifier.
+
+The withdraw function can remain hidden as a function that only the owner can call.
+
+As you can hopefully see, this removes the need to check `msg.sender` for access control purposes,
```
```suggestion
This removes the need to check `msg.sender` for access control purposes,
```

kerrywei replied:
nit: This seems to be the first time that 'msg.sender' appears in this series of tutorial. Perhaps worth to mention this is the Ethereum idiom to handle access control

joshuahannan replied:
Good point! I made that change

rheaplex commented on docs/tutorial/03-fungible-tokens.mdx:
```diff
@@ -512,22 +430,200 @@ Account storage is indexed with paths, which consist of a domain and identifier.
+
+In our example, the `Vault` resource will implement all three of these interfaces.
+The interfaces ensure that specific fields and functions are present in the resource implementation
+and that those functions meet certain conditions before and/or after execution.
```
```suggestion
and that the function arguments, fields of the resource, and any return value are in a valid state before and/or after execution.
```

rheaplex commented on docs/tutorial/03-fungible-tokens.mdx:
```diff
@@ -512,22 +430,200 @@ Account storage is indexed with paths, which consist of a domain and identifier.
+While we have made all our fields and functions public here, it is actually recommended to default to making
+fields private unless it is explicitly needed to be public.
+
+This is especially important for array and dictionary types, which can have their fields maliciously assigned to
```
```suggestion
This is especially important for array and dictionary types, which can have their contents maliciously mutated
```

rheaplex commented on docs/tutorial/03-fungible-tokens.mdx:
```diff
@@ -512,22 +430,200 @@ Account storage is indexed with paths, which consist of a domain and identifier.
+
+Now is an important time to remind you that account storage not namespaced by contract,
+meaning that path names could potentially conflict. This is why it is important
+choose unique names for your paths like we have done here so there is a very low chance
```
```suggestion
to choose unique names for your paths like we have done here so there is a very low chance
```

rheaplex commented on docs/tutorial/03-fungible-tokens.mdx:
```diff
@@ -559,8 +655,18 @@ pub fun mintTokens(amount: UFix64, recipient: Capability<&AnyResource{Receiver}>
-The function takes a capability to any resource that implements the `Receiver` interface, borrows a reference from it,
-and uses the reference to call the `deposit` function of that resource.
+The function takes a capability as an argument. 
+This syntax might be confusing to you:
```
```suggestion
This syntax might be unclear to you:
```

rheaplex reviewed (approved):
This is great! I have some suggestions but apart from adding a "to" to a sentence I don't think any are must-changes.

👍👍👍

joshuahannan commented:
Thank you everyone for the comments! I really appreciate it. I included almost every suggestion that was made. 
 @10thfloor I addressed all your comments, so can you approve the PR so I can merge it? Thanks!
